xmppserver/src/main/java/org/jivesoftware/openfire/session/LocalSession.java

```diff
@@ -519,4 +524,35 @@ public String getCipherSuiteName() {
     public final Locale getLanguage() {
         return language;
     }
+
+    /**
+     * Retrieves SoftwareVersion data. This method gives access to temporary Software Version data only. You can
+     * Please see
+     * {@link #setSoftwareVersionData(String, String)}  description for more details.
+     *
+     * @param key a <code>String</code> value of stored data ID.
+     * @return a <code>Map<String,String></code> value of data for given key.
+     * @see #setSoftwareVersionData(String, String)
+     */
+    @Override
+    public Map<String, String> getSoftwareVersion() {
+        synchronized (softwareVersionData) {
```

Comment:
When a thread-safe Map implementation is used, then there's no need to synchronize! I suggest switching from HashMap to ConcurrentHashMap and remove synchronization!